File: backend/src/main/kotlin/com/example/backend/dtos/DashboardDto.kt
package com.example.backend.dtos

import java.time.LocalDate
import java.time.LocalDateTime

data class KpiDto(
  val ca: Double,
  val encaisse: Double,
  val tickets: Long,
  val tauxRetour: Double,
  val depenses: Double,
  val sessions: SessionsDto,
  val alerts: AlertsDto,
  val caSeries: List<Double> = emptyList()
)

data class SessionsDto(
  val ouvertes: Long,
  val cloturees: Long
)

data class AlertsDto(
  val ruptures: Long,
  val perimes30j: Long
)

data class SalesMonthlyPoint(
  val mois: String,          // "YYYY-MM"
  val total: Double
)

data class CategorySales(
  val categorie: String,
  val total: Double
)

data class TopProduct(
  val nom: String,
  val qty: Long
)

data class OrderRow(
  val id: Long,
  val ref: String?,
  val fournisseur: String?,
  val montantCmd: Double?,
  val montantRecu: Double?,
  val etat: String?,
  val dateCreation: LocalDateTime?,
  val dateLivraison: LocalDateTime?
)

data class StockAlertRow(
  val produit: String,
  val quantiteRestante: Int?,
  val datePeremption: LocalDate?
)


interface CategorySalesRow { fun getCategorie(): String?; fun getTotal(): Double }

interface TopProductRow { fun getNom(): String; fun getQty(): Long }

// === Tables ===
interface OrderRowView {
  fun getId(): Long
  fun getRef(): String?
  fun getFournisseur(): String?
  fun getMontantCmd(): Double?
  fun getMontantRecu(): Double?
  fun getEtat(): String?
  fun getDateCreation(): LocalDateTime?
  fun getDateLivraison(): LocalDateTime?
}

interface StockAlertRowView {
  fun getProduit(): String
  fun getQuantiteRestante(): Int?
  fun getDatePeremption(): LocalDate?
}

File: backend_py/pharmaPython/app/schemas/dashboard_dto.py
# schemas/dashboard.py
from __future__ import annotations
from pydantic import BaseModel
from typing import Optional, List, Protocol, runtime_checkable
from datetime import datetime, date

# ---- DTOs ----

class SessionsDto(BaseModel):
  ouvertes: int
  cloturees: int


class AlertsDto(BaseModel):
  ruptures: int
  perimes30j: int


class KpiDto(BaseModel):
  ca: float
  encaisse: float
  tickets: int
  tauxRetour: float
  depenses: float
  sessions: SessionsDto
  alerts: AlertsDto
  caSeries: List[float] = []


class SalesMonthlyPoint(BaseModel):
  mois: str          # "YYYY-MM"
  total: float


class CategorySales(BaseModel):
  categorie: str
  total: float


class TopProduct(BaseModel):
  nom: str
  qty: int


class OrderRow(BaseModel):
  id: int
  ref: Optional[str] = None
  fournisseur: Optional[str] = None
  montantCmd: Optional[float] = None
  montantRecu: Optional[float] = None
  etat: Optional[str] = None
  dateCreation: Optional[datetime] = None
  dateLivraison: Optional[datetime] = None


class StockAlertRow(BaseModel):
  produit: str
  quantiteRestante: Optional[int] = None
  datePeremption: Optional[date] = None


# ---- Interfaces Kotlin → Protocols Python (facultatif)
# Utile si vous mappez des lignes SQL/ORM directement sans Pydantic.

@runtime_checkable
class CategorySalesRow(Protocol):
  def getCategorie(self) -> Optional[str]: ...
  def getTotal(self) -> float: ...


@runtime_checkable
class TopProductRow(Protocol):
  def getNom(self) -> str: ...
  def getQty(self) -> int: ...


@runtime_checkable
class OrderRowView(Protocol):
  def getId(self) -> int: ...
  def getRef(self) -> Optional[str]: ...
  def getFournisseur(self) -> Optional[str]: ...
  def getMontantCmd(self) -> Optional[float]: ...
  def getMontantRecu(self) -> Optional[float]: ...
  def getEtat(self) -> Optional[str]: ...
  def getDateCreation(self) -> Optional[datetime]: ...
  def getDateLivraison(self) -> Optional[datetime]: ...


@runtime_checkable
class StockAlertRowView(Protocol):
  def getProduit(self) -> str: ...
  def getQuantiteRestante(self) -> Optional[int]: ...
  def getDatePeremption(self) -> Optional[date]: ...
